File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/contactnotifier/comm/CarrierHelper.java
package org.elastos.trinity.runtime.contactnotifier.comm;

import android.content.Context;
import android.util.Log;

import org.elastos.carrier.AbstractCarrierHandler;
import org.elastos.carrier.Carrier;
import org.elastos.carrier.ConnectionStatus;
import org.elastos.carrier.FriendInfo;
import org.elastos.carrier.PresenceStatus;
import org.elastos.carrier.UserInfo;
import org.elastos.carrier.exceptions.CarrierException;
import org.elastos.trinity.runtime.contactnotifier.ContactNotifier;
import org.elastos.trinity.runtime.contactnotifier.OnlineStatusMode;
import org.elastos.trinity.runtime.contactnotifier.RemoteNotificationRequest;
import org.elastos.trinity.runtime.notificationmanager.NotificationManager;
import org.elastos.trinity.runtime.notificationmanager.NotificationRequest;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;

public class CarrierHelper {
    String didSessionDID;
    private ContactNotifier notifier;
    private Context context;
    Carrier carrierInstance;
    private ArrayList<CarrierCommand> commandQueue = new ArrayList<>(); // List of commands to execute. We use a queue in case we have to wait for our carrier instance to be ready (a few seconds)
    private OnCarrierEventListener onCarrierEventListener;

    public interface OnCommandExecuted {
        void onCommandExecuted(boolean succeeded, String reason);
    }

    public interface OnCarrierEventListener {
        void onFriendRequest(String did, String userId);
        void onFriendOnlineStatusChange(FriendInfo info);
        void onFriendPresenceStatusChange(FriendInfo info);
        void onRemoteNotification(String friendId, RemoteNotificationRequest remoteNotification);
    }

    public CarrierHelper(ContactNotifier notifier, String didSessionDID, Context context) throws CarrierException {
        this.notifier = notifier;
        this.context = context;
        this.didSessionDID = didSessionDID;

        initialize();
    }

    private void initialize() throws CarrierException {
        // Initial setup
        Carrier.Options options = new DefaultCarrierOptions(context.getFilesDir().getAbsolutePath()+"/contactnotifier/"+didSessionDID);

        // Create or get an our carrier instance instance
        carrierInstance = Carrier.createInstance(options, new AbstractCarrierHandler() {
            @Override
            public void onConnection(Carrier carrier, ConnectionStatus status) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier connection status: " + status);

                if(status == ConnectionStatus.Connected) {

                    // TMP TEST NOTIF
                    NotificationRequest testNotif = new NotificationRequest();
                    testNotif.key = "testkey";
                    testNotif.title = "Contact notifier is ready!";
                    try {
                        NotificationManager.getSharedInstance().sendNotification(testNotif, "org.elastos.trinity.dapp.did");
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                    // TMP TEST NOTIF END

                    // We are now connected to carrier network, we can start to send friend requests, or messages
                    checkRunQueuedCommands();
                }
            }

            @Override
            public void onFriendRequest(Carrier carrier,
                                        String userId,
                                        UserInfo info,
                                        String hello) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier received friend request. Peer UserId: " + userId);

                // First make sure this is a elastOS contact notifier plugin request, and that we understand the data
                // packaged in the hello string.
                try {
                    JSONObject invitationRequest = new JSONObject(hello);

                    String contactDID = invitationRequest.getString("did"); // Will throw exception is not present

                    Log.i(ContactNotifier.LOG_TAG, "Received friend request from DID "+contactDID+" with carrier userId: " + userId);

                    onCarrierEventListener.onFriendRequest(contactDID, userId);
                }
                catch (JSONException e) {
                    // Invitation is not understood, forget it.
                    Log.w(ContactNotifier.LOG_TAG, "Invitation received from carrier userId "+userId+" but hello string can't be understood: "+hello);
                }
            }

            @Override
            public void onFriendAdded(Carrier carrier, FriendInfo info) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier friend added. Peer UserId: " + info.getUserId());
            }

            @Override
            public void onFriendConnection(Carrier carrier,
                                           String friendId,
                                           ConnectionStatus status) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier friend connection status changed - peer UserId: " + friendId);
                Log.i(ContactNotifier.LOG_TAG, "Friend status:" + status);

                try {
                    FriendInfo info = carrier.getFriend(friendId);
                    onCarrierEventListener.onFriendOnlineStatusChange(info);
                }
                catch (CarrierException e) {
                    e.printStackTrace();
                    // Nothing
                }
            }

            @Override
            public void onFriendPresence(Carrier carrier, String friendId, PresenceStatus presence) {
                try {
                    FriendInfo info = carrier.getFriend(friendId);
                    onCarrierEventListener.onFriendPresenceStatusChange(info);
                }
                catch (CarrierException e) {
                    e.printStackTrace();
                    // Nothing
                }
            }

            @Override
            public void onFriendMessage(Carrier carrier, String from, byte[] message, boolean isOffline) {
                Log.i(ContactNotifier.LOG_TAG, "Message from userId: " + from);
                Log.i(ContactNotifier.LOG_TAG, "Message: " + new String(message));

                // Try to read this as JSON. If not json, this is an invalid command
                try {
                    JSONObject request = new JSONObject(new String(message));
                    handleReceivedMessageCommand(from, request);
                }
                catch (JSONException e) {
                    Log.i(ContactNotifier.LOG_TAG, "Invalid command for the contact notifier");
                    e.printStackTrace();
                }
            }
        });

        // Start the service
        carrierInstance.start(5000); // Start carrier. Wait N milliseconds between each check of carrier status (polling)
    }

    public void setCarrierEventListener(OnCarrierEventListener listener) {
        this.onCarrierEventListener = listener;
    }

    public String getOrCreateAddress() throws CarrierException {
        return carrierInstance.getAddress();
    }

    public void sendInvitation(String contactCarrierAddress, OnCommandExecuted completionListener) {
        queueCommand(new ContactInvitationCommand(this, contactCarrierAddress, completionListener));
    }

    public void acceptFriend(String contactCarrierUserID, OnCommandExecuted completionListener) {
        queueCommand(new AcceptFriendCommand(this, contactCarrierUserID, completionListener));
    }

    public void sendRemoteNotification(String contactCarrierUserID, RemoteNotificationRequest notificationRequest, OnCommandExecuted completionListener) {
        queueCommand(new RemoteNotificationCommand(this, contactCarrierUserID, notificationRequest, completionListener));
    }

    public void setOnlineStatusMode(OnlineStatusMode onlineStatusMode) {
        queueCommand(new SetPresenceCommand(this, notifier.onlineStatusModeToPresenceStatus(onlineStatusMode)));
    }

    public void removeFriend(String contactCarrierUserID, OnCommandExecuted completionListener) {
        queueCommand(new RemoveFriendCommand(this, contactCarrierUserID, completionListener));
    }

    public ConnectionStatus getFriendOnlineStatus(String friendId) {
        try {
            if (!carrierInstance.isReady())
                return ConnectionStatus.Disconnected;

            return carrierInstance.getFriend(friendId).getConnectionStatus();
        }
        catch (CarrierException e) {
            return ConnectionStatus.Disconnected;
        }
    }

    private void queueCommand(CarrierCommand command) {
        commandQueue.add(command);
        checkRunQueuedCommands();
    }

    /**
     * Checks if we are connected to carrier and if so, sends the queued commands.
     */
    private void checkRunQueuedCommands() {
        if (!carrierInstance.isReady())
            return;

        Iterator<CarrierCommand> it = commandQueue.iterator();
        while (it.hasNext()) {
            CarrierCommand command = it.next();
            command.executeCommand();

            // Even if the command execution fails, we remove it from the queue. We don't want to be stuck forever on a
            // corrupted command. In such case for now, we would loose the command though, which is not perfect and should be
            // improved to be more robust.
            it.remove();
        }
    }

    private void handleReceivedMessageCommand(String friendId, JSONObject request) {
        if (!request.has("command")) {
            Log.w(ContactNotifier.LOG_TAG, "Command received as JSON, but no command field inside");
            return;
        }

        try {
            String command = request.getString("command");

            switch (command) {
                case "remotenotification":
                    handleReceivedRemoteNotification(friendId, request);
                    break;
                default:
                    Log.w(ContactNotifier.LOG_TAG, "Unknown command: "+command);
            }
        }
        catch (JSONException e) {
            e.printStackTrace();
            Log.w(ContactNotifier.LOG_TAG, "Invalid remote command received");
        }
    }

    private void handleReceivedRemoteNotification(String friendId, JSONObject request) {
        if (!request.has("key") || !request.has("title")) {
            Log.w(ContactNotifier.LOG_TAG, "Invalid remote notification command received: missing mandatory fields");
            return;
        }

        RemoteNotificationRequest remoteNotification = RemoteNotificationRequest.fromJSONObject(request);
        if (remoteNotification == null) {
            // Couldn't parse as a proper notification.
            Log.w(ContactNotifier.LOG_TAG, "Invalid remote notification command received: format not understood");
            return;
        }

        onCarrierEventListener.onRemoteNotification(friendId, remoteNotification);
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/ContactNotifier/comm/CarrierHelper.swift
import ElastosCarrierSDK

protocol OnCarrierEventListener {
    func onFriendRequest(_ did: String, _ userId: String)
    func onFriendOnlineStatusChange(_ info: FriendInfo)
    func onFriendPresenceStatusChange(_ info: FriendInfo)
    func onRemoteNotification(_ friendId: String, _ remoteNotification: RemoteNotificationRequest)
}

public class CarrierHelper {
    var didSessionDID: String
    private var notifier: ContactNotifier
    var carrierInstance: Carrier? = nil
    private var commandQueue = Array<CarrierCommand>() // List of commands to execute. We use a queue in case we have to wait for our carrier instance to be ready (a few seconds)
    private var onCarrierEventListener: OnCarrierEventListener? = nil
    private var carrierDelegate: CarrierDelegate? = nil

    public typealias onCommandExecuted = (_ succeeded: Bool, _ reason: String?) -> Void

    public init(notifier: ContactNotifier, didSessionDID: String) throws {
        self.notifier = notifier
        self.didSessionDID = didSessionDID
        
        try initialize()
    }

    private func initialize() throws {
        // Initial setup
        let options = DefaultCarrierOptions.createOptions(didSessionDID: didSessionDID)
        
        class CarrierHandler : CarrierDelegate {
            let helper: CarrierHelper
            
            init(helper: CarrierHelper) {
                self.helper = helper
            }
            
            func connectionStatusDidChange(_ carrier: Carrier, _ status: CarrierConnectionStatus) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier connection status: \(status)")

                if(status == .Connected) {
                    // We are now connected to carrier network, we can start to send friend requests, or messages
                    helper.checkRunQueuedCommands()
                }
            }
            
            func didReceiveFriendRequest(_ carrier: Carrier, _ userId: String, _ userInfo: CarrierUserInfo, _ hello: String) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier received friend request. Peer UserId: \(userId)");

                // First make sure this is a elastOS contact notifier plugin request, and that we understand the data
                // packaged in the hello string.
                if let invitationRequest = hello.toDict() { // JSON stirng to JSON object
                    if let contactDID = invitationRequest["did"] as? String {
                        Log.i(ContactNotifier.LOG_TAG, "Received friend request from DID \(String(describing: contactDID)) with carrier userId: " + userId);

                        helper.onCarrierEventListener?.onFriendRequest(contactDID, userId)
                    }
                    else {
                        Log.w(ContactNotifier.LOG_TAG, "Invitation received from carrier userId \(userId) but no contact DID is given.")
                    }
                }
                else {
                    // Invitation is not understood, forget it.
                    Log.w(ContactNotifier.LOG_TAG, "Invitation received from carrier userId \(userId) but hello string can't be understood: \(hello)")
                }
            }
            
            func didReceiveFriendInviteRequest(_ carrier: Carrier, _ from: String, _ data: String) {
                Log.i(ContactNotifier.LOG_TAG, "Did receive friend invite request from: \(from)")
            }
            
            func newFriendAdded(_ carrier: Carrier, _ info: CarrierFriendInfo) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier friend added. Peer UserId: \(String(describing: info.userId))")
            }
            
            func friendConnectionDidChange(_ carrier: Carrier, _ friendId: String, _ status: CarrierConnectionStatus) {
                Log.i(ContactNotifier.LOG_TAG, "Carrier friend connection status changed - peer UserId: \(friendId)")
                Log.i(ContactNotifier.LOG_TAG, "Friend status: \(status)")

                if let info = try? carrier.getFriendInfo(friendId) {
                    helper.onCarrierEventListener?.onFriendOnlineStatusChange(info)
                }
            }
            
            func friendPresenceDidChange(_ carrier: Carrier, _ friendId: String, _ newPresence: CarrierPresenceStatus) {
                if let info = try? carrier.getFriendInfo(friendId) {
                    helper.onCarrierEventListener?.onFriendPresenceStatusChange(info)
                }
            }
            
            func didReceiveFriendMessage(_ carrier: Carrier, _ from: String, _ data: Data, _ isOffline: Bool) {
                let dataAsStr = String(data: data, encoding: .utf8)
                
                Log.i(ContactNotifier.LOG_TAG, "Message from userId: \(from)")
                Log.i(ContactNotifier.LOG_TAG, "Message: \(String(describing: dataAsStr))")

                // Try to read this as JSON. If not json, this is an invalid command
                if let request = dataAsStr?.toDict() {
                    handleReceivedMessageCommand(friendId: from, request: request)
                }
                else {
                    Log.i(ContactNotifier.LOG_TAG, "Received friend message but unable to read it as json")
                }
            }
            
            private func handleReceivedMessageCommand(friendId: String, request: Dictionary<String, Any>) {
                if !request.keys.contains("command") {
                    Log.w(ContactNotifier.LOG_TAG, "Command received as JSON, but no command field inside")
                    return
                }

                let command = request["command"] as? String

                switch (command) {
                    case "remotenotification":
                        handleReceivedRemoteNotification(friendId: friendId, request: request)
                        break;
                    default:
                        Log.w(ContactNotifier.LOG_TAG, "Unknown command: \(String(describing: command))")
                }
            }

            private func handleReceivedRemoteNotification(friendId: String, request: Dictionary<String, Any>) {
                if !request.keys.contains("key") || !request.keys.contains("title") {
                    Log.w(ContactNotifier.LOG_TAG, "Invalid remote notification command received: missing mandatory fields")
                    return
                }

                guard let remoteNotification = RemoteNotificationRequest.fromJSONObject(request) else {
                    // Couldn't parse as a proper notification.
                    Log.w(ContactNotifier.LOG_TAG, "Invalid remote notification command received: format not understood")
                    return
                }

                helper.onCarrierEventListener?.onRemoteNotification(friendId, remoteNotification)
            }
        }

        carrierDelegate = CarrierHandler(helper: self) // Retain a reference on carrier delegate as carrier FWK does not, and this leads to a crash.
        carrierInstance = try Carrier.createInstance(options: options, delegate: carrierDelegate!)

        // Start the service
        try carrierInstance!.start(iterateInterval: 5000) // Start carrier. Wait N milliseconds between each check of carrier status (polling)
    }
    
    func setCarrierEventListener(_ listener: OnCarrierEventListener) {
        self.onCarrierEventListener = listener
    }

    public func getOrCreateAddress() throws -> String{
        return carrierInstance!.getAddress()
    }

    public func sendInvitation(contactCarrierAddress: String, completionListener: @escaping onCommandExecuted) {
        queueCommand(ContactInvitationCommand(helper: self, contactCarrierAddress: contactCarrierAddress, completionListener: completionListener))
    }

    public func acceptFriend(contactCarrierUserID: String, completionListener: @escaping onCommandExecuted) {
        queueCommand(AcceptFriendCommand(helper: self, contactCarrierUserID: contactCarrierUserID, completionListener: completionListener))
    }

    public func sendRemoteNotification(contactCarrierUserID: String, notificationRequest: RemoteNotificationRequest, completionListener: @escaping onCommandExecuted) {
        queueCommand(RemoteNotificationCommand(helper: self, contactCarrierUserID: contactCarrierUserID, notificationRequest: notificationRequest, completionListener: completionListener))
    }

    public func setOnlineStatusMode(_ onlineStatusMode: OnlineStatusMode) {
        queueCommand(SetPresenceCommand(helper: self, status: notifier.onlineStatusModeToPresenceStatus(onlineStatusMode)))
    }

    public func removeFriend(contactCarrierUserID: String, completionListener: @escaping onCommandExecuted) {
        queueCommand(RemoveFriendCommand(helper: self, contactCarrierUserID: contactCarrierUserID, completionListener: completionListener))
    }

    public func getFriendOnlineStatus(friendId: String) -> CarrierConnectionStatus {
        do {
            if (!carrierInstance!.isReady()) {
                return .Disconnected
            }

            return try carrierInstance!.getFriendInfo(friendId).status
        }
        catch {
            return .Disconnected
        }
    }

    private func queueCommand(_ command: CarrierCommand) {
        commandQueue.append(command)
        checkRunQueuedCommands()
    }

    /**
     * Checks if we are connected to carrier and if so, sends the queued commands.
     */
    private func checkRunQueuedCommands() {
        guard carrierInstance!.isReady() else {
            return
        }

        while commandQueue.count > 0 {
            if let command = commandQueue.first {
                command.executeCommand()
                
                // Even if the command execution fails, we remove it from the queue. We don't want to be stuck forever on a
                // corrupted command. In such case for now, we would loose the command though, which is not perfect and should be
                // improved to be more robust.
                commandQueue.removeFirst()
            }
        }
    }
}
